lib/Storage/MongoDbTimeSeriesStore.php

```diff
@@ -120,137 +127,231 @@ public function setDependencies($log, $date, $layoutFactory = null, $campaignFac
-                    $statData[$k]['widgetName'] = $widgetName;
+                    return;
                 }
             }
 
-            // Display name
-            try {
-                $display = $this->displayFactory->getById($stat['displayId']); //TODO what if you dont find the displayId
-            } catch (NotFoundException $error) {
-                // Display not found, ignore and log the stat
-                $this->log->error('Display not found. Display Id: '. $stat['displayId'] .',Layout Id: '. $stat['layoutId']
-                    .', FromDT: '.$statData[$k]['start'].', ToDt: '.$statData[$k]['end'].', Type: '.$stat['type']
-                    .', Duration: '.$stat['duration'] .', Count '.$stat['count']);
-
-                // unset($statData[$k]);
-                continue;
-            }
+            // SET widgetName
+            $statData['widgetName'] = $widgetName;
 
-            $statData[$k]['displayName'] = $display->display;
+        }
+
+        // Display name
+        try {
+            $display = $this->displayFactory->getById($statData['displayId']);
+        } catch (NotFoundException $error) {
+            // Display not found, ignore and log the stat
+            $this->log->error('Display not found. Display Id: '. $statData['displayId'] .',Layout Id: '. $statData['layoutId']
+                .', FromDT: '.$statData['start'].', ToDt: '.$statData['end'].', Type: '.$statData['type']
+                .', Duration: '.$statData['duration'] .', Count '.$statData['count']);
+
+            return;
+        }
 
-            // Layout data
-            $layoutName = null;
-            $layoutTags = null;
+        $statData['displayName'] = $display->display;
 
-            if ($stat['type'] != 'event') {
+        // Layout data
+        $layoutName = null;
+        $layoutTags = null;
+
+        // For a type "event" we have layoutid 0 so is campaignId
+        // otherwise we should try and resolve the campaignId
+        $campaignId = 0;
+        if ($statData['type'] != 'event') {
+            if (array_key_exists($statData['layoutId'], $this->layoutIds)) {
+
+                $campaignId = $this->layoutIds[$statData['layoutId']]['campaignId'];
+                $layoutName = $this->layoutIds[$statData['layoutId']]['name'];
+                $layoutTags = $this->layoutIds[$statData['layoutId']]['tags'];
+            } else {
 
                 try {
-                    $layout = $this->layoutFactory->getByParentCampaignId($stat['campaignId']);
 
-                    $this->log->debug('Found layout : '. $stat['layoutId']);
+                    // Get the layout campaignId - this can give us a campaignId of a layoutId which id was replaced when draft to published
+                    $campaignId = $this->layoutFactory->getCampaignIdFromLayoutHistory($statData['layoutId']);
+
+                    // Get the layout by campaignId
+                    $layout = $this->layoutFactory->getByParentCampaignId($campaignId);
```

Comment:
I think you could have a `layoutFactory->getByLayoutHistory($layoutId)` method which combined these two queries into 1?